pkg/inframetadata/internal/hostmap/hostmap.go

```diff
@@ -0,0 +1,89 @@
+package hostmap
+
+import (
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/multierr"
+
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/gohai"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/payload"
+)
+
+// HostMap stores a map from hostnames to host metadata payloads.
+// Host metadata payloads can be updated with the information from a given resource.
+// At any time, the internal map can be extracted and the map will be cleared out.
+type HostMap struct {
+	// mutex for the host map.
+	mutex sync.Mutex
+	// hosts map
+	hosts map[string]payload.HostMetadata
+}
+
+// New creates a new HostMap
+func New() (*HostMap, error) {
+	return &HostMap{
+		mutex: sync.Mutex{},
+		hosts: make(map[string]payload.HostMetadata),
+	}, nil
+}
+
+// Update the information about a given host by providing a resource.
+// The function reports:
+//   - Whether the information about the `host` has changed
+//   - Any non-fatal errors that may have occurred during the update
+//
+// Partial modifications will still be applied even with non-fatal errors.
+func (m *HostMap) Update(host string, res pcommon.Resource) (changed bool, err error) {
+	hm := payload.HostMetadata{
+		Flavor:  "otelcol-contrib",
+		Meta:    &payload.Meta{},
+		Tags:    &payload.HostTags{},
+		Payload: gohai.NewEmpty(),
+	}
+	updater := newUpdater(res)
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	var existingHost bool
+	if oldHM, ok := m.hosts[host]; ok {
+		existingHost = true
+		hm = oldHM
+	}
+
+	updater.FromValue(&hm.InternalHostname, host)
+
+	// Meta section
+	updater.FromFunc(&hm.Meta.InstanceID,
+		func(m pcommon.Map) (string, bool, error) {
+			return "", false, fmt.Errorf("not implemented: InstanceID")
+		},
+	)
+	updater.FromFunc(&hm.Meta.EC2Hostname,
+		func(m pcommon.Map) (string, bool, error) {
+			return "", false, fmt.Errorf("not implemented: EC2Hostname")
+		},
+	)
+	updater.FromValue(&hm.Meta.Hostname, host)
+
+	// Gohai - Platform
+	updater.FromValue(hm.Gohai.Gohai.Platform.(map[string]*string)["hostname"], host)
+	for field, attribute := range platformAttributesMap {
+		updater.FromAttributeToMap(hm.Gohai.Gohai.Platform.(map[string]string), field, attribute)
+	}
```

Comment:
This part needs some inline documentation and clarification. It is too hard to understand what's going on here. I suspect we can make some improvements on this end.

I feel like the `updater` could be simplified by simply taking an `hm` and a `host` and do everything under the hood, but I don't understand what's going on here that well yet to be able to suggest improvements. Maybe later on in a follow up review?

Reply:
I got rid of the updater, it's more verbose but I think it was worth it